faiss/impl/AdditiveQuantizer.cpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// -*- c++ -*-
+
+#include <faiss/impl/AdditiveQuantizer.h>
+#include <faiss/impl/FaissAssert.h>
+
+#include <cstddef>
+#include <cstdio>
+#include <cstring>
+#include <memory>
+#include <random>
+
+#include <algorithm>
+
+#include <faiss/utils/distances.h>
+#include <faiss/utils/hamming.h> // BitstringWriter
+#include <faiss/utils/utils.h>
+
+namespace {
+
+// c and a and b can overlap
+void fvec_add(size_t d, const float* a, const float* b, float* c) {
+    for (size_t i = 0; i < d; i++) {
+        c[i] = a[i] + b[i];
+    }
+}
+
+} // namespace
+
+namespace faiss {
+
+void AdditiveQuantizer::set_derived_values() {
+    tot_bits = 0;
+    is_byte_aligned = true;
+    codebook_offsets.resize(M + 1, 0);
+    for (int i = 0; i < M; i++) {
+        int nbit = nbits[i];
+        size_t k = 1 << nbit;
+        codebook_offsets[i + 1] = codebook_offsets[i] + k;
+        tot_bits += nbit;
+        if (nbit % 8 != 0) {
+            is_byte_aligned = false;
+        }
+    }
+    // convert bits to bytes
+    code_size = (tot_bits + 7) / 8;
+}
+
+void AdditiveQuantizer::train(size_t n, const float* x) {
```

Comment:
Please declare this as =0 in the .h file (then the compiler will make sure people don't instantiate the class).